homeassistant/components/guardian/__init__.py

```diff
@@ -82,105 +58,96 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
         )
     )
     if unload_ok:
-        hass.data[DOMAIN][DATA_CLIENT].pop(entry.entry_id)
+        hass.data[DOMAIN][DATA_COORDINATOR].pop(entry.entry_id)
 
     return unload_ok
 
 
-class Guardian:
-    """Define a class to communicate with the Guardian device."""
+class Guardian(DataUpdateCoordinator):
+    """Define a class to communicate with a Guardian valve controller.
+
+    Normally, we'd use a single DataUpdateCoordinator class for each API call; however,
```

Comment:
What if we share make one coordinator per API call and share an asyncio.Lock between them that we await before fetching data? Then we could do away with the custom api interest registration that is already handled by the coordinator.

Reply:
Ahhhh, very intriguing. I like that. Will experiment.

Reply:
I like that, but the only challenge there is that it’s slightly wasteful: each coordinator will open a connection to the device, perform a single API request, then close the connection (whereas what I have here will perform all API requests during a single connection). I suppose it comes down the whether simpler code wins our over (probably small) performance... 🤔

Reply:
@MartinHjelmare Added this in 3dc1ea108930. No noticeable performance degradation. 👍 Let me know what you think?